File: app/src/lib/storage.ts
import { invoke } from '@tauri-apps/api/core';
import type { AnalysisResult, Classification } from './analysis';
import type { GameReview } from './reviewEngine';
import type { ParsedGame } from './parsePgn';
import { bookExitPly, openingFrom } from './openings';
import { terminationFrom } from './termination';

export interface ReviewSummary {
  id: string;
  savedAt: number;
  white: string;
  black: string;
  result: string;
  date: string;
  moveCount: number;
  whiteAccuracy: number;
  blackAccuracy: number;
  whiteCounts: Partial<Record<Classification, number>>;
  blackCounts: Partial<Record<Classification, number>>;
  whiteMotifs: Record<string, number>;
  blackMotifs: Record<string, number>;
  opening?: string | null;
  eco?: string | null;
  bookExitPly?: number | null;
  termination?: string | null;
}

export interface StoredReview {
  summary: ReviewSummary;
  pgn: string;
  analysis: AnalysisResult[];
  depth: number;
  multiPv: number;
}

export function reviewId(pgn: string): string {
  let hash = 2166136261;
  for (let i = 0; i < pgn.length; i++) {
    hash ^= pgn.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return `${(hash >>> 0).toString(16)}-${pgn.length.toString(16)}`;
}

function tally(
  review: GameReview,
  color: 'w' | 'b',
): { counts: Record<string, number>; motifs: Record<string, number> } {
  const counts: Record<string, number> = {};
  const motifs: Record<string, number> = {};
  for (const move of review.moves) {
    if (move.color !== color) continue;
    counts[move.classification] = (counts[move.classification] ?? 0) + 1;
    const motif = move.explanation.motif;
    if (motif && motif !== 'positive' && motif !== 'evaluation' && motif !== 'none') {
      motifs[motif] = (motifs[motif] ?? 0) + 1;
    }
  }
  return { counts, motifs };
}

export function buildSummary(
  id: string,
  game: ParsedGame,
  review: GameReview,
): ReviewSummary {
  const white = tally(review, 'w');
  const black = tally(review, 'b');
  const opening = openingFrom(game.headers);
  return {
    opening: opening.name,
    eco: opening.eco,
    bookExitPly: bookExitPly(game),
    termination: terminationFrom(game.headers),
    id,
    savedAt: Date.now(),
    white: game.headers.White ?? 'Unknown',
    black: game.headers.Black ?? 'Unknown',
    result: game.headers.Result ?? '*',
    date: game.headers.Date ?? game.headers.UTCDate ?? '',
    moveCount: review.moves.length,
    whiteAccuracy: review.whiteAccuracy,
    blackAccuracy: review.blackAccuracy,
    whiteCounts: white.counts,
    blackCounts: black.counts,
    whiteMotifs: white.motifs,
    blackMotifs: black.motifs,
  };
}

export async function listReviews(): Promise<ReviewSummary[]> {
  return invoke<ReviewSummary[]>('list_reviews');
}

export async function loadReview(id: string): Promise<StoredReview> {
  return invoke<StoredReview>('load_review', { id });
}

export async function saveReview(review: StoredReview): Promise<void> {
  return invoke('save_review', { review });
}

export async function deleteReview(id: string): Promise<void> {
  return invoke('delete_review', { id });
}

File: app/src-tauri/src/storage.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use tauri::Manager;

use crate::AnalysisResult;

const MAX_STORED_REVIEWS: usize = 500;

#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ReviewSummary {
    pub id: String,
    pub saved_at: i64,
    pub white: String,
    pub black: String,
    pub result: String,
    pub date: String,
    pub move_count: usize,
    pub white_accuracy: f64,
    pub black_accuracy: f64,
    #[serde(default)]
    pub white_counts: HashMap<String, u32>,
    #[serde(default)]
    pub black_counts: HashMap<String, u32>,
    #[serde(default)]
    pub white_motifs: HashMap<String, u32>,
    #[serde(default)]
    pub black_motifs: HashMap<String, u32>,
    #[serde(default)]
    pub opening: Option<String>,
    #[serde(default)]
    pub eco: Option<String>,
    #[serde(default)]
    pub book_exit_ply: Option<usize>,
    #[serde(default)]
    pub termination: Option<String>,
}

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct StoredReview {
    pub summary: ReviewSummary,
    pub pgn: String,
    pub analysis: Vec<AnalysisResult>,
    pub depth: u32,
    pub multi_pv: u32,
}

#[derive(Default)]
pub struct StorageLock(Mutex<()>);

fn reviews_dir(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Couldn't locate the app data directory: {e}"))?
        .join("reviews");
    fs::create_dir_all(&dir)
        .map_err(|e| format!("Couldn't create {}: {e}", dir.display()))?;
    Ok(dir)
}

fn index_path(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    Ok(reviews_dir(app)?.join("index.json"))
}

fn is_valid_review_id(id: &str) -> bool {
    !id.is_empty() && id.chars().all(|c| c.is_ascii_alphanumeric() || c == '-')
}

fn game_path(app: &tauri::AppHandle, id: &str) -> Result<PathBuf, String> {
    if !is_valid_review_id(id) {
        return Err(format!("Refusing to use {id:?} as a review id"));
    }
    Ok(reviews_dir(app)?.join(format!("{id}.json")))
}

fn write_atomic(path: &Path, contents: &str) -> Result<(), String> {
    let tmp = path.with_extension("json.tmp");
    fs::write(&tmp, contents).map_err(|e| format!("Couldn't write {}: {e}", tmp.display()))?;
    fs::rename(&tmp, path).map_err(|e| format!("Couldn't finalize {}: {e}", path.display()))?;
    Ok(())
}

fn read_index(app: &tauri::AppHandle) -> Vec<ReviewSummary> {
    let Ok(path) = index_path(app) else {
        return Vec::new();
    };
    let Ok(raw) = fs::read_to_string(path) else {
        return Vec::new();
    };
    serde_json::from_str(&raw).unwrap_or_default()
}

#[tauri::command]
pub fn list_reviews(app: tauri::AppHandle) -> Vec<ReviewSummary> {
    let mut index = read_index(&app);
    index.sort_by(|a, b| b.saved_at.cmp(&a.saved_at));
    index
}

#[tauri::command]
pub fn load_review(app: tauri::AppHandle, id: String) -> Result<StoredReview, String> {
    let path = game_path(&app, &id)?;
    let raw = fs::read_to_string(&path)
        .map_err(|_| "That saved review is no longer available.".to_string())?;
    serde_json::from_str(&raw).map_err(|e| format!("That saved review couldn't be read: {e}"))
}

#[tauri::command]
pub fn save_review(
    app: tauri::AppHandle,
    lock: tauri::State<StorageLock>,
    review: StoredReview,
) -> Result<(), String> {
    let _guard = lock.0.lock().map_err(|_| "Storage lock poisoned".to_string())?;

    let path = game_path(&app, &review.summary.id)?;
    let body = serde_json::to_string(&review).map_err(|e| e.to_string())?;
    write_atomic(&path, &body)?;

    let mut index = read_index(&app);
    index.retain(|s| s.id != review.summary.id);
    index.push(review.summary);
    index.sort_by(|a, b| b.saved_at.cmp(&a.saved_at));

    if index.len() > MAX_STORED_REVIEWS {
        for pruned in index.split_off(MAX_STORED_REVIEWS) {
            if let Ok(p) = game_path(&app, &pruned.id) {
                let _ = fs::remove_file(p);
            }
        }
    }

    let index_body = serde_json::to_string(&index).map_err(|e| e.to_string())?;
    write_atomic(&index_path(&app)?, &index_body)
}

#[tauri::command]
pub fn delete_review(
    app: tauri::AppHandle,
    lock: tauri::State<StorageLock>,
    id: String,
) -> Result<(), String> {
    let _guard = lock.0.lock().map_err(|_| "Storage lock poisoned".to_string())?;

    let path = game_path(&app, &id)?;
    let _ = fs::remove_file(path);

    let mut index = read_index(&app);
    index.retain(|s| s.id != id);
    let index_body = serde_json::to_string(&index).map_err(|e| e.to_string())?;
    write_atomic(&index_path(&app)?, &index_body)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_ids_that_could_escape_the_reviews_directory() {
        for id in [
            "../secret",
            "a/b",
            "..",
            "",
            "with space",
            "semi;colon",
            "back\\slash",
            "null\0byte",
            "dot.dot",
        ] {
            assert!(
                !is_valid_review_id(id),
                "{id:?} should be rejected as a review id"
            );
        }
    }

    #[test]
    fn accepts_content_hash_ids() {
        for id in ["a1b2c3d4", "0", "abc-123", "deadbeefcafe0123"] {
            assert!(
                is_valid_review_id(id),
                "{id:?} should be accepted as a review id"
            );
        }
    }
}
